### DISCUSSION TITLE
How to run gifs as animation using flash memory for BT815 ?

### DISCUSSION THREAD

**Discussion by theSwiftGuy** [RudolphRiedel/FT800-FT813#123] [General]
Hi Rudolph,

Long time no see. 
I need some help from you regarding GIFs .
I want to add 2 gifs into my XMC project. I created a BLOB file that contains them.
I am using BT815 variant, so I cannot use inflate method like I did for bitmaps(PNG).
From what I read, I understand that I can use gifs in my project only if I use flash memory. Luckily, I have 64 Mb memory on my display.

Looking into some examples, I saw that I have to use some flash functions in order to load and run the GIFs as animations.
Until now, I tried only to initiate flash memory using this code( in order to see that my flash memory is responsive)

uint8_t EVE_init_flash(void)
{
	uint8_t timeout = 0;
	uint8_t status;

      /* should be 0x02 - FLASH_STATUS_BASIC, power-up is done and the attached flash is detected */
	status = EVE_memRead8(REG_FLASH_STATUS); 

	while(status == 0) /* FLASH_STATUS_INIT - we are somehow still in init, give it a litte more time, this should never happen */
	{
		status = EVE_memRead8(REG_FLASH_STATUS);
		osDelay(1);
		timeout++;
		if(timeout > 100) /* 100ms and still in init, lets call quits now and exit with an error */
		{
			return 0;
		}
	}

	if(status == 1) /* FLASH_STATUS_DETACHED - no flash was found during init, no flash present or the detection failed, but have hope and let the BT81x have another try */
	{
		EVE_cmd_dl(CMD_FLASHATTACH);
    		EVE_cmd_execute();
		status = EVE_memRead8(REG_FLASH_STATUS);
		if(status != 2) /* still not in FLASH_STATUS_BASIC, time to give up */
		{
			return 0;
		}
	}

	if(status == 2) /* FLASH_STATUS_BASIC - flash detected and ready for action, lets move it up to FLASH_STATUS_FULL */
	{
		uint32_t result;
		result = EVE_cmd_flashfast();
                status = EVE_memRead8(REG_FLASH_STATUS);
		if(result == 0) /* cmd_flashfast was successful */
		{
			return 1;
		}
		else
		{
			return 0;
		}
	}

	if(status == 3) /* FLASH_STATUS_FULL - we are already there, why has this function been called? */
	{
		return 1;
	}

	return 0;
}

As a result for variables „result” and „status” I got values for result=57346 and status=0x02. What 57346 means? It is that value correct? It is the size of my flash memory? From what I know, I have a 64mb size.
 
I was expected perhaps to see status 0x03 after I call cmd_flashfast method, but the status remain the same. I don’t know why, I don’t know if I missed something important . Beside this,  I don’t find REG_FLASH_FULL in the manual or in the code . Which is the value for this ?

I saw that functions like cmd_flashfast and cmd_flashupdate are also need it. 

Also, EVE_CMD_ANIMFRAME must be used after the init flash is done, and the program is running.

Could you help me with this? It is not clear for me how I must use flash related methods, in order to make gifs work as animations.
An example it will be very helpful.
PS: Did you received your XMC board, and Keil license ?

Best regards!
Camil

**Comment by RudolphRiedel** [comment]
I have not really tried animations and I am not sure what I would want to use these for.
On the one side animations have the potential to save storage space by splitting images into tiles and not all of these need to be updated every time.
But on the other side this eats up the display list.

Name              : bicycle.anim.flash
Resolution        : 400x300
ASTC Block        : 8x8
ASTC Preset       : thorough
Tile Size         : auto (2x2)
Number Of Frames  : 13
Memory Usage      : 108288 (bytes)
Display List Usage: 8380
    Each Frame    : 642, 642, 642, 651 (Max), 645, 643, 648, 648, 649, 641 (Min), 641 (Min), 644, 644

Tile Size         : 6x6
Number Of Frames  : 13
Memory Usage      : 149760 (bytes)
Display List Usage: 1365
    Each Frame    : 105, 105, 105, 108 (Max), 105, 104 (Min), 104 (Min), 104 (Min), 105, 104 (Min), 106, 105, 105

Tile Size         : not apply
Number Of Frames  : 13
Memory Usage      : 396160 (bytes)
Display List Usage: 130
    Each Frame    : 10 (same for all frames)

In this example I could just store the 13 frames in flash and use two buffers of 30.000 bytes each in RAM_G, copy one frame from flash while displaying the other.
This might be even working good enough without douple buffering.

>As a result for variables „result” and „status” I got values for result=57346 and status=0x02. What 57346 means?

57346 is 0xE002 which means "no header detected in sector 0 - is flash blank?" - so yes, the flash is probably empty.
And 0x02 is FLASH_STATUS_BASIC - so at least the flash has been detected and initialized.

>Beside this, I don’t find REG_FLASH_FULL in the manual or in the code . Which is the value for this ?

Interesting, I see this in the programming guide but there really is no REG_FLASH_FULL register.
Looks like someone messed up the example code and really meant to write this:
while (FLASH_STATUS_FULL!= rd8(REG_FLASH_STATUS));

My EVE_init_flash() function does the complete initialization though and reports back what happened trying to do so.

>PS: Did you received your XMC board, and Keil license ?

Yes and no, I have a "XMC4700 Relax Lite Kit" here with me but gave up on getting a licence for the Keil software.
Somehow I am stacking eval boards here without really getting anywhere, I do currently have at least four more.

**Comment by RudolphRiedel** [comment]
I am playing half the day now with the EK-TM4C1294XL board and it feels like I am not getting anywhere.
I do have Keil µVision 5.39, Arm Keil Studio Pack (MDK v6) and TIs Code Composer Studio 12.7.1 installed now.
And so far the development tools are a nightmare.
At least it looks like I do have a community licence for µVision now.

I managed to setup a project for the TM4C1294 in Arm Keil Studio in VSCode and it compiles.
But I can not flash the software to the board, it looks like the onboard Debugger is not compatible with the MDK v6 - but it does not say so.
Keil µVision 5.39 can not build that project.
Setting up a new project in Keil µVision 5.39 left me with an empty shell with nothing in it to edit or compile, this is not actually a project, there is no startup code, nothing.
Edit: and the examples do not build with µVision 5.39 as these require ARM compiler V5 but there is only V6 included

Setting up a new project in Code Composer Studio worked better and the project compiles - but I can not figure out how to even try to flash the software to the board - and this is not Keil anyways.
There is TI TIVA for PlatformIO but it might not even support SPI and is definately not Keil.
I could try to figure out how to work with Segger Embedded Studio - not Keil either.

Experience is 0/10 so far.
